src/routes/Project/actions/TrackActionHandler.js

```diff
@@ -47,101 +82,168 @@ const ACTION_HANDLERS = {
-            newClip.endTime = timeClip.endTime + clipLength + 1;
-            newClip.startTime = timeClip.endTime + 1;
-          }
-
-          // and recurse
-          adjustForCollisions(newClip, track);
-        }
-      }
-    };
-
-    adjustForCollisions(newClip, track);
-
     if (_.isEqual(oldClip, newClip)) return state;
 
     let newTracks;
-    if (oldClip.track !== newClip.track) {
+    const collision = isCollision(newClip, state[newClip.track]);
+    if (oldClip.track !== newClip.track && !collision) {
+      if (oldClip.ghostClip) newClip.ghostClip = false;
+
       let oldTrack = { ...state[oldClip.track] };
-      delete oldTrack[oldClip.startTime];
-      delete oldTrack[oldClip.endTime];
       delete oldTrack.clips[id];
-      oldTrack.edgeClip = findEdgeClip(oldTrack);
+      oldTrack.edgeClip = findEdgeClip(oldTrack.clips);
 
       newTracks = {
         ...state,
         [oldClip.track]: oldTrack,
         [newClip.track]: {
           ...state[newClip.track],
-          [newClip.startTime]: id,
-          [newClip.endTime]: id,
           clips: {
             ...state[newClip.track].clips,
             [id]: newClip
           }
         }
       };
 
-      newTracks[newClip.track].edgeClip = findEdgeClip(newTracks[newClip.track]);
+      newTracks[newClip.track].edgeClip = findEdgeClip(newTracks[newClip.track].clips);
     } else {
+      if (collision) {
+        newClip.track = oldClip.track;
+      }
+
+      if (!oldClip.ghostClip) {
+        adjustForCollisions(newClip, state[oldClip.track], isMovingRight);
+      } else if (oldClip.ghostClip && !isCollision(newClip, state[oldClip.track])) {
+        newClip.ghostClip = false;
+      }
+
       newTracks = {
         ...state,
-        [newClip.track]: {
-          ...state[newClip.track],
-          [newClip.startTime]: id,
-          [newClip.endTime]: id,
+        [oldClip.track]: {
+          ...state[oldClip.track],
           clips: {
-            ...state[newClip.track].clips,
+            ...state[oldClip.track].clips,
             [id]: newClip
           }
         }
       };
 
-      delete newTracks[newClip.track][oldClip.startTime];
-      delete newTracks[newClip.track][oldClip.endTime];
-      newTracks[newClip.track].edgeClip = findEdgeClip(newTracks[newClip.track]);
+      newTracks[oldClip.track].edgeClip = findEdgeClip(newTracks[oldClip.track].clips);
     }
 
     return newTracks;
+  },
+
+  [ProjectConstants.DELETE_CLIP]: (state, action) => {
+    const clip = action.payload;
+    let newClips = { ...state[clip.track].clips };
+    delete newClips[clip.id];
+
+    return {
+      ...state,
+      [clip.track]: {
+        ...state[clip.track],
+        clips: newClips,
+        edgeClip: findEdgeClip(newClips)
+      }
+    };
   }
 };
 
 export default ACTION_HANDLERS;
 
 /* HELPERS */
-function findEdgeClip(track) {
+function findEdgeClip(clips) {
   let max;
-  for (let time in track) {
+  let maxClip;
+  for (let clipId in clips) {
```

Comment:
instead of all of this, just use:
`_.maxBy(clips, 'endTime')`
https://lodash.com/docs/4.16.4#maxBy
lodash is 🔑